presto-main/src/main/java/com/facebook/presto/sql/planner/plan/StatisticAggregationsDescriptor.java

```diff
@@ -78,37 +124,76 @@ public StatisticAggregationsDescriptor(
+    private static <K, V1, V2> Map<K, V2> map(Map<K, V1> input, Function<V1, V2> mapper)
     {
         return input.entrySet()
                 .stream()
-                .collect(toImmutableMap(entry -> mapper.apply(entry.getKey()), Map.Entry::getValue));
+                .collect(toImmutableMap(Map.Entry::getKey, entry -> mapper.apply(entry.getValue())));
     }
 
     public static class Builder<T>
     {
-        private final ImmutableMap.Builder<T, String> grouping = ImmutableMap.builder();
-        private final ImmutableMap.Builder<T, TableStatisticType> tableStatistics = ImmutableMap.builder();
-        private final ImmutableMap.Builder<T, ColumnStatisticMetadata> columnStatistics = ImmutableMap.builder();
+        private final ImmutableMap.Builder<String, T> grouping = ImmutableMap.builder();
+        private final ImmutableMap.Builder<TableStatisticType, T> tableStatistics = ImmutableMap.builder();
+        private final ImmutableMap.Builder<ColumnStatisticMetadata, T> columnStatistics = ImmutableMap.builder();
 
-        public void addGrouping(T key, String column)
+        public void addGrouping(String column, T key)
         {
-            grouping.put(key, column);
+            grouping.put(column, key);
         }
 
-        public void addTableStatistic(T key, TableStatisticType type)
+        public void addTableStatistic(TableStatisticType type, T key)
         {
-            tableStatistics.put(key, type);
+            tableStatistics.put(type, key);
         }
 
-        public void addColumnStatistic(T key, ColumnStatisticMetadata statisticMetadata)
+        public void addColumnStatistic(ColumnStatisticMetadata statisticMetadata, T key)
         {
-            columnStatistics.put(key, statisticMetadata);
+            columnStatistics.put(statisticMetadata, key);
         }
 
         public StatisticAggregationsDescriptor<T> build()
         {
             return new StatisticAggregationsDescriptor<>(grouping.build(), tableStatistics.build(), columnStatistics.build());
         }
     }
+
+    public static class ColumnStatisticMetadataKeySerializer
+            extends JsonSerializer<ColumnStatisticMetadata>
+    {
+        @Override
+        public void serialize(ColumnStatisticMetadata value, JsonGenerator gen, SerializerProvider serializers)
+                throws IOException
+        {
+            verify(value != null, "value is null");
+            gen.writeFieldName(serialize(value));
+        }
+
+        @VisibleForTesting
+        public static String serialize(ColumnStatisticMetadata value)
+        {
+            return value.getStatisticType().name() + ":" + value.getColumnName();
+        }
+    }
+
+    public static class ColumnStatisticMetadataKeyDeserializer
+            extends KeyDeserializer
+    {
+        @Override
+        public ColumnStatisticMetadata deserializeKey(String key, DeserializationContext ctxt)
+        {
+            verify(key != null, "key is null");
```

Comment:
requireNonNull?

Reply:
This is the situation that should never happen. The `ColumnStatisticMetadata` key is never supposed to be null. `verify` is just a runtime `assert` that is supposed to be showing that this check is an explicit assertion about the possible value. I don't have a strong opinion though.

Reply:
changed to `requireNonNull`

Reply:
Technically it can, since this an "API" of the worker. It's handling data that originates from a different process/machine, so there's no hard guarantee that it's not null (other than by virtue of running the same version of the code)